Dataset is Taken from kaggle website

__This Dataset is Pokemon dataset where we have Pokemon names and their characteristics like attack,defence speed etc.
So First we are going to select the best pokemons for the fight and then we are going to see what are the factors influencing a Pokemon Attack. And lastly we will see if the Pokemon is Legendary or Not.__

#### A) Load the data and show first four rows

In [2]:
pokemon <- read.csv('datasets_2619_4359_pokemon.csv')

In [3]:
head(pokemon,4)

X. Name          Type.1 Type.2 HP Attack Defense Sp..Atk Sp..Def Speed
1 1  Bulbasaur     Grass  Poison 45  49     49      65      65     45   
2 2  Ivysaur       Grass  Poison 60  62     63      80      80     60   
3 3  Venusaur      Grass  Poison 80  82     83     100     100     80   
4 4  Mega Venusaur Grass  Poison 80 100    123     122     120     80   
  Generation Legendary
1 1          False    
2 1          False    
3 1          False    
4 1          False

Here we can see there is no need of the first column because its just row numbers
So lets remove it 

In [4]:
library('dplyr')

Warning message:
"package 'dplyr' was built under R version 3.6.3"
Attaching package: 'dplyr'

The following objects are masked from 'package:stats':

    filter, lag

The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union



In [5]:
pokemon <- pokemon %>% select(-1) 

In [6]:
head(pokemon,4)

Name          Type.1 Type.2 HP Attack Defense Sp..Atk Sp..Def Speed
1 Bulbasaur     Grass  Poison 45  49     49      65      65     45   
2 Ivysaur       Grass  Poison 60  62     63      80      80     60   
3 Venusaur      Grass  Poison 80  82     83     100     100     80   
4 Mega Venusaur Grass  Poison 80 100    123     122     120     80   
  Generation Legendary
1 1          False    
2 1          False    
3 1          False    
4 1          False

In [7]:
nrow(pokemon)

[1] 800

In [8]:
ncol(pokemon)

[1] 11

Renaming some columns for better understanding

In [9]:
"Primary_Type" -> colnames(pokemon)[2]
"Secondary_Type" -> colnames(pokemon)[3]
"Health_Points" -> colnames(pokemon)[4]
"Special_Attack" -> colnames(pokemon)[7]
"Special_Defence" -> colnames(pokemon)[8]

In [10]:
head(pokemon,4)

Name          Primary_Type Secondary_Type Health_Points Attack Defense
1 Bulbasaur     Grass        Poison         45             49     49    
2 Ivysaur       Grass        Poison         60             62     63    
3 Venusaur      Grass        Poison         80             82     83    
4 Mega Venusaur Grass        Poison         80            100    123    
  Special_Attack Special_Defence Speed Generation Legendary
1  65             65             45    1          False    
2  80             80             60    1          False    
3 100            100             80    1          False    
4 122            120             80    1          False

In [11]:
str(pokemon)

'data.frame':	800 obs. of  11 variables:
 $ Name           : Factor w/ 800 levels "","Abomasnow",..: 70 309 746 454 89 90 88 420 421 669 ...
 $ Primary_Type   : Factor w/ 18 levels "Bug","Dark","Dragon",..: 10 10 10 10 7 7 7 7 7 18 ...
 $ Secondary_Type : Factor w/ 19 levels "","Bug","Dark",..: 15 15 15 15 1 1 9 4 9 1 ...
 $ Health_Points  : int  45 60 80 80 39 58 78 78 78 44 ...
 $ Attack         : int  49 62 82 100 52 64 84 130 104 48 ...
 $ Defense        : int  49 63 83 123 43 58 78 111 78 65 ...
 $ Special_Attack : int  65 80 100 122 60 80 109 130 159 50 ...
 $ Special_Defence: int  65 80 100 120 50 65 85 85 115 64 ...
 $ Speed          : int  45 60 80 80 65 80 100 100 100 43 ...
 $ Generation     : int  1 1 1 1 1 1 1 1 1 1 ...
 $ Legendary      : Factor w/ 2 levels "False","True": 1 1 1 1 1 1 1 1 1 1 ...


## (DATA WRANGLING)
###  Choose one best Pokemon from each type Grass,Water and Fire . 

In [12]:
table(pokemon$Primary_Type)


     Bug     Dark   Dragon Electric    Fairy Fighting     Fire   Flying 
      69       31       32       44       17       27       52        4 
   Ghost    Grass   Ground      Ice   Normal   Poison  Psychic     Rock 
      32       70       32       24       98       28       57       44 
   Steel    Water 
      27      112 

__Select a Grass Pokemon which is Poisonous and has the Maximum Speed__

In [13]:
Grass_pokemon <- pokemon %>% filter(Primary_Type=="Grass")

In [14]:
Grass_pokemon

Name               Primary_Type Secondary_Type Health_Points Attack Defense
1   Bulbasaur          Grass        Poison         45             49     49    
2   Ivysaur            Grass        Poison         60             62     63    
3   Venusaur           Grass        Poison         80             82     83    
4   Mega Venusaur      Grass        Poison         80            100    123    
5   Oddish             Grass        Poison         45             50     55    
6   Gloom              Grass        Poison         60             65     70    
7   Vileplume          Grass        Poison         75             80     85    
8   Bellsprout         Grass        Poison         50             75     35    
9   Weepinbell         Grass        Poison         65             90     50    
10  Victreebel         Grass        Poison         80            105     65    
11  Exeggcute          Grass        Psychic        60             40     80    
12  Exeggutor          Grass        Psychic        95             95     85    
13  Tangela            Grass                       65             55    115    
14  Chikorita          Grass                       45             49     65    
15  Bayleef            Grass                       60             62     80    
16  Meganium           Grass                       80             82    100    
17  Bellossom          Grass                       75             80     95    
18  Hoppip             Grass        Flying         35             35     40    
19  Skiploom           Grass        Flying         55             45     50    
20  Jumpluff           Grass        Flying         75             55     70    
21  Sunkern            Grass                       30             30     30    
22  Sunflora           Grass                       75             75     55    
23  Treecko            Grass                       40             45     35    
24  Grovyle            Grass                       50             65     45    
25  Sceptile           Grass                       70             85     65    
26  Mega Sceptile      Grass        Dragon         70            110     75    
27  Seedot             Grass                       40             40     50    
28  Nuzleaf            Grass        Dark           70             70     40    
29  Shiftry            Grass        Dark           90            100     60    
30  Shroomish          Grass                       60             40     60    
... ...                ...          ...            ...           ...    ...    
41  Cherubi            Grass                        45            35     45    
42  Cherrim            Grass                        70            60     70    
43  Carnivine          Grass                        74           100     72    
44  Snover             Grass        Ice             60            62     50    
45  Abomasnow          Grass        Ice             90            92     75    
46  Mega Abomasnow     Grass        Ice             90           132    105    
47  Tangrowth          Grass                       100           100    125    
48  Leafeon            Grass                        65           110    130    
49  Shaymin Land Forme Grass                       100           100    100    
50  Shaymin Sky Forme  Grass        Flying         100           103     75    
51  Snivy              Grass                        45            45     55    
52  Servine            Grass                        60            60     75    
53  Serperior          Grass                        75            75     95    
54  Pansage            Grass                        50            53     48    
55  Simisage           Grass                        75            98     63    
56  Cottonee           Grass        Fairy           40            27     60    
57  Whimsicott         Grass        Fairy           60            67     85    
58  Petilil            Grass                        45            35     50    
59  

In [15]:
nrow(Grass_pokemon)

[1] 70

In [16]:
Grass_Poison_pokemon <- Grass_pokemon %>% filter(Secondary_Type=="Poison")

In [17]:
Grass_Poison_pokemon

Name          Primary_Type Secondary_Type Health_Points Attack Defense
1  Bulbasaur     Grass        Poison          45            49     49    
2  Ivysaur       Grass        Poison          60            62     63    
3  Venusaur      Grass        Poison          80            82     83    
4  Mega Venusaur Grass        Poison          80           100    123    
5  Oddish        Grass        Poison          45            50     55    
6  Gloom         Grass        Poison          60            65     70    
7  Vileplume     Grass        Poison          75            80     85    
8  Bellsprout    Grass        Poison          50            75     35    
9  Weepinbell    Grass        Poison          65            90     50    
10 Victreebel    Grass        Poison          80           105     65    
11 Roselia       Grass        Poison          50            60     45    
12 Budew         Grass        Poison          40            30     35    
13 Roserade      Grass        Poison          60            70     65    
14 Foongus       Grass        Poison          69            55     45    
15 Amoonguss     Grass        Poison         114            85     70    
   Special_Attack Special_Defence Speed Generation Legendary
1   65             65             45    1          False    
2   80             80             60    1          False    
3  100            100             80    1          False    
4  122            120             80    1          False    
5   75             65             30    1          False    
6   85             75             40    1          False    
7  110             90             50    1          False    
8   70             30             40    1          False    
9   85             45             55    1          False    
10 100             70             70    1          False    
11 100             80             65    3          False    
12  50             70             55    4          False    
13 125            105             90    4          False    
14  55             55             15    5          False    
15  85             80             30    5          False

In [18]:
nrow(Grass_Poison_pokemon)

[1] 15

In [19]:
range(Grass_Poison_pokemon$Speed)

[1] 15 90

So there is one pokemon having the maximum speed as 90. we will select that pokemon.

In [20]:
My_Grass_Pokemon <- Grass_Poison_pokemon %>% filter(Speed==90)

In [21]:
My_Grass_Pokemon

Name     Primary_Type Secondary_Type Health_Points Attack Defense
1 Roserade Grass        Poison         60            70     65     
  Special_Attack Special_Defence Speed Generation Legendary
1 125            105             90    4          False

Here we selected our Grass Pokemon that is Roserade.

__Now we will select our Water Pokemon__

In [22]:
Water_pokemon <- pokemon %>% filter(Primary_Type=="Water")

In [23]:
Water_pokemon

Name                  Primary_Type Secondary_Type Health_Points Attack
1   Squirtle              Water                        44            48   
2   Wartortle             Water                        59            63   
3   Blastoise             Water                        79            83   
4   Mega Blastoise        Water                        79           103   
5   Psyduck               Water                        50            52   
6   Golduck               Water                        80            82   
7   Poliwag               Water                        40            50   
8   Poliwhirl             Water                        65            65   
9   Poliwrath             Water        Fighting        90            95   
10  Tentacool             Water        Poison          40            40   
11  Tentacruel            Water        Poison          80            70   
12  Slowpoke              Water        Psychic         90            65   
13  Slowbro               Water        Psychic         95            75   
14  Mega Slowbro          Water        Psychic         95            75   
15  Seel                  Water                        65            45   
16  Dewgong               Water        Ice             90            70   
17  Shellder              Water                        30            65   
18  Cloyster              Water        Ice             50            95   
19  Krabby                Water                        30           105   
20  Kingler               Water                        55           130   
21  Horsea                Water                        30            40   
22  Seadra                Water                        55            65   
23  Goldeen               Water                        45            67   
24  Seaking               Water                        80            92   
25  Staryu                Water                        30            45   
26  Starmie               Water        Psychic         60            75   
27  Magikarp              Water                        20            10   
28  Gyarados              Water        Flying          95           125   
29  Mega Gyarados         Water        Dark            95           155   
30  Lapras                Water        Ice            130            85   
... ...                   ...          ...            ...           ...   
83  Gastrodon             Water        Ground         111            83   
84  Finneon               Water                        49            49   
85  Lumineon              Water                        69            69   
86  Mantyke               Water        Flying          45            20   
87  Palkia                Water        Dragon          90           120   
88  Phione                Water                        80            80   
89  Manaphy               Water                       100           100   
90  Oshawott              Water                        55            55   
91  Dewott                Water                        75            75   
92  Samurott              Water                        95           100   
93  Panpour               Water                        50            53   
94  Simipour              Water                        75            98   
95  Tympole               Water                        50            50   
96  Palpitoad             Water        Ground          75            65   
97  Seismitoad            Water        Ground         105            95   
98  Basculin              Water                        70            92   
99  Tirtouga              Water        Rock            54            78   
100 Carracosta            Water        Rock            74           108   
101 Ducklett              Water        Flying          62            44   
102 Swanna                Water        Flying          75            87   
103 Frillish              Water        Ghost           55            40   
104 Jellicent             Wat

In [24]:
nrow(Water_pokemon)

[1] 112

In [25]:
Water_Psychic_pokemon <- Water_pokemon %>% filter(Secondary_Type=="Psychic")

In [26]:
Water_Psychic_pokemon

Name         Primary_Type Secondary_Type Health_Points Attack Defense
1 Slowpoke     Water        Psychic        90            65      65    
2 Slowbro      Water        Psychic        95            75     110    
3 Mega Slowbro Water        Psychic        95            75     180    
4 Starmie      Water        Psychic        60            75      85    
5 Slowking     Water        Psychic        95            75      80    
  Special_Attack Special_Defence Speed Generation Legendary
1  40             40              15   1          False    
2 100             80              30   1          False    
3 130             80              30   1          False    
4 100             85             115   1          False    
5 100            110              30   2          False

In [27]:
range(Water_Psychic_pokemon$Defense)

[1]  65 180

So here the Maximum Defense value is 180 we will select that pokemon

In [28]:
My_Water_Pokemon <- Water_Psychic_pokemon %>% filter(Defense==180)

In [29]:
My_Water_Pokemon

Name         Primary_Type Secondary_Type Health_Points Attack Defense
1 Mega Slowbro Water        Psychic        95            75     180    
  Special_Attack Special_Defence Speed Generation Legendary
1 130            80              30    1          False

So our Water Pokemon is Mega Slowbro.

__Now we will select our last Fire Pokemon__

In [30]:
Fire_pokemon <- pokemon %>% filter(Primary_Type=="Fire")

In [31]:
Fire_pokemon

Name                     Primary_Type Secondary_Type Health_Points Attack
1  Charmander               Fire                         39            52   
2  Charmeleon               Fire                         58            64   
3  Charizard                Fire         Flying          78            84   
4  Mega Charizard X         Fire         Dragon          78           130   
5  Mega Charizard Y         Fire         Flying          78           104   
6  Vulpix                   Fire                         38            41   
7  Ninetales                Fire                         73            76   
8  Growlithe                Fire                         55            70   
9  Arcanine                 Fire                         90           110   
10 Ponyta                   Fire                         50            85   
11 Rapidash                 Fire                         65           100   
12 Magmar                   Fire                         65            95   
13 Flareon                  Fire                         65           130   
14 Moltres                  Fire         Flying          90           100   
15 Cyndaquil                Fire                         39            52   
16 Quilava                  Fire                         58            64   
17 Typhlosion               Fire                         78            84   
18 Slugma                   Fire                         40            40   
19 Magcargo                 Fire         Rock            50            50   
20 Magby                    Fire                         45            75   
21 Entei                    Fire                        115           115   
22 Ho-oh                    Fire         Flying         106           130   
23 Torchic                  Fire                         45            60   
24 Combusken                Fire         Fighting        60            85   
25 Blaziken                 Fire         Fighting        80           120   
26 Mega Blaziken            Fire         Fighting        80           160   
27 Numel                    Fire         Ground          60            60   
28 Camerupt                 Fire         Ground          70           100   
29 Mega Camerupt            Fire         Ground          70           120   
30 Torkoal                  Fire                         70            85   
31 Chimchar                 Fire                         44            58   
32 Monferno                 Fire         Fighting        64            78   
33 Infernape                Fire         Fighting        76           104   
34 Magmortar                Fire                         75            95   
35 Heatran                  Fire         Steel           91            90   
36 Tepig                    Fire                         65            63   
37 Pignite                  Fire         Fighting        90            93   
38 Emboar                   Fire         Fighting       110           123   
39 Pansear                  Fire                         50            53   
40 Simisear                 Fire                         75            98   
41 Darumaka                 Fire                         70            90   
42 Darmanitan Standard Mode Fire                        105           140   
43 Darmanitan Zen Mode      Fire         Psychic        105            30   
44 Heatmor                  Fire                         85            97   
45 Fennekin                 Fire                         40            45   
46 Braixen                  Fire                         59            59   
47 Delphox                  Fire         Psychic         75            69   
48 Fletchinder              Fire         Flying          62            73   
49 Talonflame               Fire         Flying          78            81   
50 Litleo                   Fire         Normal          62            50   
51 Pyroar                   Fire         Normal          86            68   

In [32]:
Fire_Fighting_pokemon <- Fire_pokemon %>% filter(Secondary_Type=="Fighting")

In [33]:
Fire_Fighting_pokemon

Name          Primary_Type Secondary_Type Health_Points Attack Defense
1 Combusken     Fire         Fighting        60            85    60     
2 Blaziken      Fire         Fighting        80           120    70     
3 Mega Blaziken Fire         Fighting        80           160    80     
4 Monferno      Fire         Fighting        64            78    52     
5 Infernape     Fire         Fighting        76           104    71     
6 Pignite       Fire         Fighting        90            93    55     
7 Emboar        Fire         Fighting       110           123    65     
  Special_Attack Special_Defence Speed Generation Legendary
1  85            60               55   3          False    
2 110            70               80   3          False    
3 130            80              100   3          False    
4  78            52               81   4          False    
5 104            71              108   4          False    
6  70            55               55   5          False    
7 100            65               65   5          False

In [34]:
range(Fire_Fighting_pokemon$Attack)

[1]  78 160

In [35]:
My_Fire_Pokemon <- Fire_Fighting_pokemon %>% filter(Attack==160)

In [36]:
My_Fire_Pokemon

Name          Primary_Type Secondary_Type Health_Points Attack Defense
1 Mega Blaziken Fire         Fighting       80            160    80     
  Special_Attack Special_Defence Speed Generation Legendary
1 130            80              100   3          False

We selected our final Fire Pokemon that is Mega Blaziken.

We can now bind all our three pokemon in My pokemon 

and see our result in one place

In [37]:
My_Pokemons <- rbind(My_Grass_Pokemon,My_Water_Pokemon,My_Fire_Pokemon)

In [38]:
My_Pokemons

Name          Primary_Type Secondary_Type Health_Points Attack Defense
1 Roserade      Grass        Poison         60             70     65    
2 Mega Slowbro  Water        Psychic        95             75    180    
3 Mega Blaziken Fire         Fighting       80            160     80    
  Special_Attack Special_Defence Speed Generation Legendary
1 125            105              90   4          False    
2 130             80              30   1          False    
3 130             80             100   3          False

## (REGRESSION)
### Understand what are the factors influencing a Pokemon's attack! 

So here we are using __Linear Regression__ 

In [39]:
# Splitting data into train and test sets
#library(MASS)
library(caTools)

Warning message:
"package 'caTools' was built under R version 3.6.3"

In [40]:
split_index <- sample.split(pokemon$Attack,SplitRatio =0.65)

In [41]:
train1 <- subset(pokemon,split_index==T)

In [42]:
test1 <- subset(pokemon,split_index==F)

#### Building Model 1 
#### Attack VS Defense

In [43]:
mod_regress <- lm(Attack~Defense,data = train1)

In [44]:
result_regress <- predict(mod_regress,test1)

In [45]:
Final_data <- cbind(Actual = test1$Attack, predicted=result_regress)

In [46]:
Final_data <- as.data.frame(Final_data)

In [47]:
Final_data

Actual predicted
1    49     67.28371
3    82     83.63845
4   100    102.87932
6    64     71.61290
8   130     97.10706
14   30     60.54940
21   45     62.95451
23   80     79.79027
24   80     82.19538
29   60     64.87860
32   90     70.16984
33   75     84.60049
34  100     96.62604
35   47     68.72677
38   57     62.95451
41   45     66.80269
43   41     62.95451
47   45     60.54940
49   50     70.16984
56   55     55.73919
60   52     66.80269
61   82     81.23334
63  105     72.57495
64   70     65.35962
66   50     62.95451
68   95     89.41071
74  100     77.38517
76   75     60.54940
82   95     99.03115
91   65     70.16984
... ...    ...      
698  85     70.16984
700  90    105.76545
702  90     78.34721
708 150    101.43625
715  72     87.00560
716  77     80.75232
719  61     74.98006
726  63     68.72677
727  95     75.94210
729  56     80.75232
732  81     77.86619
735  52     67.76473
742 100     73.53699
745  80     72.57495
746  48     69.68882
749  80     91.81582
751 150     67.76473
752  50    115.86691
764  73     86.04356
765  38     59.58736
767  89     80.75232
769  59     67.76473
778  80     87.48662
784  66     77.38517
786  85    102.39830
788 100    102.39830
790 117    132.22165
794 131     89.41071
799 160     72.57495
800 110    101.43625

Now we will find the error in prediction

In [48]:
error <- (Final_data$Actual - Final_data$predicted)

In [49]:
Final_data <- cbind(Final_data,error)

In [50]:
Final_data

Actual predicted error      
1    49     67.28371 -18.2837085
3    82     83.63845  -1.6384490
4   100    102.87932  -2.8793201
6    64     71.61290  -7.6129045
8   130     97.10706  32.8929412
14   30     60.54940 -30.5494036
21   45     62.95451 -17.9545125
23   80     79.79027   0.2097253
24   80     82.19538  -2.1953836
29   60     64.87860  -4.8785996
32   90     70.16984  19.8301608
33   75     84.60049  -9.6004925
34  100     96.62604   3.3739630
35   47     68.72677 -21.7267739
38   57     62.95451  -5.9545125
41   45     66.80269 -21.8026867
43   41     62.95451 -21.9545125
47   45     60.54940 -15.5494036
49   50     70.16984 -20.1698392
56   55     55.73919  -0.7391858
60   52     66.80269 -14.8026867
61   82     81.23334   0.7666599
63  105     72.57495  32.4250519
64   70     65.35962   4.6403786
66   50     62.95451 -12.9545125
68   95     89.41071   5.5892897
74  100     77.38517  22.6148341
76   75     60.54940  14.4505964
82   95     99.03115  -4.0311459
91   65     70.16984  -5.1698392
... ...    ...       ...        
698  85     70.16984  14.830161 
700  90    105.76545 -15.765451 
702  90     78.34721  11.652791 
708 150    101.43625  48.563745 
715  72     87.00560 -15.005601 
716  77     80.75232  -3.752318 
719  61     74.98006 -13.980057 
726  63     68.72677  -5.726774 
727  95     75.94210  19.057899 
729  56     80.75232 -24.752318 
732  81     77.86619   3.133812 
735  52     67.76473 -15.764730 
742 100     73.53699  26.463008 
745  80     72.57495   7.425052 
746  48     69.68882 -21.688817 
749  80     91.81582 -11.815819 
751 150     67.76473  82.235270 
752  50    115.86691 -65.866908 
764  73     86.04356 -13.043558 
765  38     59.58736 -21.587360 
767  89     80.75232   8.247682 
769  59     67.76473  -8.764730 
778  80     87.48662  -7.486623 
784  66     77.38517 -11.385166 
786  85    102.39830 -17.398298 
788 100    102.39830  -2.398298 
790 117    132.22165 -15.221649 
794 131     89.41071  41.589290 
799 160     72.57495  87.425052 
800 110    101.43625   8.563745

Now we will find the aggregate error in prediction

We can do that by using root mean square error

In [51]:
rmse1 <- sqrt(mean(Final_data$error^2))

In [52]:
rmse1

[1] 28.85576

28 is the root mean square error

__We checked how Defense influence the Attack of the pokemon__

__Now we will see how Defense, Speed and Health points influence the Attack of the pokemon__

In [53]:
mod_regress2 <- lm(Attack~Defense+Speed+Health_Points, data=train1)

In [54]:
result_regress2 <- predict(mod_regress2,test1)

In [55]:
Final_data2 <- cbind(Actual = test1$Attack, Predicted = result_regress2)

In [56]:
Final_data2 <- as.data.frame(Final_data2)

In [57]:
Final_data2

Actual Predicted
1    49     51.55948
3    82     91.08791
4   100    106.89622
6    64     73.28386
8   130    109.13922
14   30     46.02658
21   45     50.43995
23   80     97.09792
24   80    106.77980
29   60     49.83461
32   90     84.37736
33   75     65.66137
34  100     94.17826
35   47     54.80569
38   57     50.28925
41   45     56.31574
43   41     53.18723
47   45     48.07862
49   50     48.15134
56   55     48.73297
60   52     56.81808
61   82     91.03834
63  105     82.37489
64   70     59.35979
66   50     63.53989
68   95     95.57924
74  100     72.46499
76   75     45.90099
82   95     77.39202
91   65     62.23134
... ...    ...      
698  85     63.31187
700  90    124.01758
702  90    101.49075
708 150    116.76703
715  72    108.60448
716  77     99.77310
719  61     59.14769
726  63     76.02189
727  95     98.06547
729  56     89.74696
732  81    103.34852
735  52     81.51369
742 100     93.65259
745  80     88.67369
746  48     68.52023
749  80     65.16510
751 150     63.13671
752  50    102.65747
764  73     81.73124
765  38     54.50829
767  89     68.46345
769  59     63.86560
778  80     84.03908
784  66     63.36013
786  85    104.81718
788 100     98.28429
790 117    116.37132
794 131    119.71903
799 160     81.99814
800 110    101.85767

Find error

In [58]:
error2 <- (Final_data2$Actual - Final_data2$Predicted)

In [59]:
Final_data2 <- cbind(Final_data2,error2)

In [60]:
Final_data2

Actual Predicted error2     
1    49     51.55948  -2.5594835
3    82     91.08791  -9.0879148
4   100    106.89622  -6.8962195
6    64     73.28386  -9.2838639
8   130    109.13922  20.8607768
14   30     46.02658 -16.0265768
21   45     50.43995  -5.4399529
23   80     97.09792 -17.0979188
24   80    106.77980 -26.7798030
29   60     49.83461  10.1653863
32   90     84.37736   5.6226388
33   75     65.66137   9.3386265
34  100     94.17826   5.8217415
35   47     54.80569  -7.8056919
38   57     50.28925   6.7107481
41   45     56.31574 -11.3157397
43   41     53.18723 -12.1872327
47   45     48.07862  -3.0786225
49   50     48.15134   1.8486554
56   55     48.73297   6.2670257
60   52     56.81808  -4.8180763
61   82     91.03834  -9.0383383
63  105     82.37489  22.6251079
64   70     59.35979  10.6402076
66   50     63.53989 -13.5398913
68   95     95.57924  -0.5792379
74  100     72.46499  27.5350149
76   75     45.90099  29.0990073
82   95     77.39202  17.6079822
91   65     62.23134   2.7686579
... ...    ...       ...        
698  85     63.31187  21.6881315
700  90    124.01758 -34.0175800
702  90    101.49075 -11.4907459
708 150    116.76703  33.2329708
715  72    108.60448 -36.6044830
716  77     99.77310 -22.7731017
719  61     59.14769   1.8523106
726  63     76.02189 -13.0218855
727  95     98.06547  -3.0654660
729  56     89.74696 -33.7469619
732  81    103.34852 -22.3485186
735  52     81.51369 -29.5136942
742 100     93.65259   6.3474073
745  80     88.67369  -8.6736930
746  48     68.52023 -20.5202277
749  80     65.16510  14.8348973
751 150     63.13671  86.8632922
752  50    102.65747 -52.6574694
764  73     81.73124  -8.7312352
765  38     54.50829 -16.5082938
767  89     68.46345  20.5365452
769  59     63.86560  -4.8655986
778  80     84.03908  -4.0390783
784  66     63.36013   2.6398691
786  85    104.81718 -19.8171787
788 100     98.28429   1.7157107
790 117    116.37132   0.6286837
794 131    119.71903  11.2809680
799 160     81.99814  78.0018603
800 110    101.85767   8.1423265

In [61]:
# Find the aggregate error by implementing rmse

In [62]:
rmse2 <- sqrt(mean(Final_data2$error^2))

In [63]:
rmse2

[1] 25.49142

In [64]:
rmse1

[1] 28.85576

__Conclusion :__

__second model is better than first because rmse2 is lesser as compared to rmse1__

## (CLASSIFICATION)
### Implement Classification algorithm to find out if the Pokemon is Legendary or Not? 

In [65]:
Split_values <- sample.split(pokemon$Legendary,SplitRatio = 0.65)

In [66]:
train2 <- subset(pokemon,Split_values == T)

In [67]:
test2 <- subset(pokemon, Split_values == F)

In [68]:
nrow(train2)

[1] 520

In [69]:
nrow(test2)

[1] 280

In [70]:
# Decision Tree
library(rpart)

Warning message:
"package 'rpart' was built under R version 3.6.3"

__Building model 1__

In [71]:
mod1 <- rpart(Legendary~. ,data = train2)

In [72]:
result1 <- predict(mod1, test2,type = "class")

In [73]:
table(test2$Legendary,result1)

       result1
        False True
  False   255    2
  True     16    7

In [74]:
#install.packages('caret')
library(caret)

Warning message:
"package 'caret' was built under R version 3.6.3"Loading required package: lattice
Warning message:
"package 'lattice' was built under R version 3.6.3"Loading required package: ggplot2
Warning message:
"package 'ggplot2' was built under R version 3.6.3"

In [75]:
confusionMatrix(table(test2$Legendary,result1))

Confusion Matrix and Statistics

       result1
        False True
  False   255    2
  True     16    7
                                          
               Accuracy : 0.9357          
                 95% CI : (0.9003, 0.9615)
    No Information Rate : 0.9679          
    P-Value [Acc > NIR] : 0.997973        
                                          
                  Kappa : 0.4103          
                                          
 Mcnemar's Test P-Value : 0.002183        
                                          
            Sensitivity : 0.9410          
            Specificity : 0.7778          
         Pos Pred Value : 0.9922          
         Neg Pred Value : 0.3043          
             Prevalence : 0.9679          
         Detection Rate : 0.9107          
   Detection Prevalence : 0.9179          
      Balanced Accuracy : 0.8594          
                                          
       'Positive' Class : False           
                                   

__Building model 2__

In [76]:
mod2 <- rpart(Legendary~ Defense+Speed+Attack,data = train2)

In [77]:
result2 <- predict(mod2,test2,type="class")

In [78]:
table(test2$Legendary,result2)

       result2
        False True
  False   248    9
  True     12   11

In [79]:
confusionMatrix(table(test2$Legendary,result2))

Confusion Matrix and Statistics

       result2
        False True
  False   248    9
  True     12   11
                                         
               Accuracy : 0.925          
                 95% CI : (0.8876, 0.953)
    No Information Rate : 0.9286         
    P-Value [Acc > NIR] : 0.6469         
                                         
                  Kappa : 0.4712         
                                         
 Mcnemar's Test P-Value : 0.6625         
                                         
            Sensitivity : 0.9538         
            Specificity : 0.5500         
         Pos Pred Value : 0.9650         
         Neg Pred Value : 0.4783         
             Prevalence : 0.9286         
         Detection Rate : 0.8857         
   Detection Prevalence : 0.9179         
      Balanced Accuracy : 0.7519         
                                         
       'Positive' Class : False          
                                         

So here Accuracy of Model 1 is 93% and Accuracy of Model 2 is 92%

Model 1 > Model 2

__Conclusion : Model 1 is Better Model.__


Note: model performance can vary after re-run or with some minor changes